### PR TITLE
Centralized Portnr Code and Configuration

### PR DESCRIPTION
**Pull request by chhex**
See Jira IT-34794. 

Basically two files have been added: 

1. The Portnr Calculation and Configuration Code , see https://github.com/apgsga-it/apg-gradle-properties/blob/IT-34794/common/portnr.gradle

2. A build.gradle file with a Tasks which list all Services/ Targets and their Portnummer for testing and querying purposes : https://github.com/apgsga-it/apg-gradle-properties/blob/IT-34794/common/build.gradle

The file supportedServices.json has been deleted. 

I have commented details in the changed files : https://github.com/apgsga-it/apg-gradle-properties/pull/2/files

Open Points: 

- I am not quite sure about the location of the code
- Complete Configuration 
- The calculation of the additional Ports for a Service

### CODE REVIEW

File: common/build.gradle [hunk 1]
```diff
@@ -0,0 +1,7 @@
+apply from: 'portnr.gradle'
```
Comment (chhex) [inline common/build.gradle]: Test Build, which lists all port numbers per Service / Target

File: common/portnr.gradle [hunk 2]
```diff
@@ -0,0 +1,49 @@
+ext.portNr = [
```
Comment (chhex) [inline common/portnr.gradle]: The actual Portnumber Calculation and Configuration as file.  A Gradle extra Property , which is a Map of names and anonymous closures.

File: common/portnr.gradle [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        serviceNames:{
```
Comment (chhex) [inline common/portnr.gradle]: This clojure serviceNames returns the a Map of serviceName as Key and Portnummer Part as Value. New services a configured here

File: common/portnr.gradle [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+            -> return ['jadas':1, 'digiflex-jadas':2,'digiflex-web-it21':3,'digiflex-web-sa':4, 'echoservice':5]
+        },
+        envDesignators: { ->
```
Comment (chhex) [inline common/portnr.gradle]: This clojure envDesignators  returns a Map of Target System Environment Character and the corresponding number in the Portnummer. 
Open: other environments like Production

File: common/portnr.gradle [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                return ['T':3, 'E':4]
+        },
+        list:{ ->
```
Comment (chhex) [inline common/portnr.gradle]: This clojure list calls the partnr calc clojure for all service names and a set of Targets. 
Open: Targets

File: common/portnr.gradle [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                def TARGETS = ["CHEI212", "CHEI211", "CHTI212", "CHTI211"]
+                println SERVICE_NAMES.toString()
+                def names = SERVICE_NAMES.keySet() as List
+                names.each { name ->
+                    TARGETS.each {
+                            println "Calcuated Port Service: ${name} and Target : ${it} is " + ext.portNr.calc.call(it,name)
+                    }
+                }
+        },
+        calc:{ target, serviceName ->
```
Comment (chhex) [inline common/portnr.gradle]: The actuall Portnumber calculation clojure , whch calls servicesNames and envDesignators for the configuration.  Open: see comments

File: common/portnr.gradle [hunk 7]
```diff
@@ -0,0 +1,49 @@
+ext.portNr = [
+        serviceNames:{
+            -> return ['jadas':1, 'digiflex-jadas':2,'digiflex-web-it21':3,'digiflex-web-sa':4, 'echoservice':5]
+        },
+        envDesignators: { ->
+                return ['T':3, 'E':4]
+        },
```
Comment (chhex) [inline common/portnr.gradle]: This clojure list calls the partnr calc clojure for all service names and a set of Targets.
Open: Targets

File: common/portnr.gradle [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                def SERVICE_NAMES = ext.portNr.serviceNames.call()
+                def TARGETS = ["CHEI212", "CHEI211", "CHTI212", "CHTI211"]
+                println SERVICE_NAMES.toString()
+                def names = SERVICE_NAMES.keySet() as List
+                names.each { name ->
+                    TARGETS.each {
+                            println "Calcuated Port Service: ${name} and Target : ${it} is " + ext.portNr.calc.call(it,name)
+                    }
+                }
+        },
```
Comment (chhex) [inline common/portnr.gradle]: The actual calculation, which calls serviceNames , envDesignators resp. 
Open : see TODO ' s in the code.

File: common/portnr.gradle [hunk 9]
```diff
@@ -0,0 +1,49 @@
+ext.portNr = [
+        serviceNames:{
+            -> return ['jadas':1, 'digiflex-jadas':2,'digiflex-web-it21':3,'digiflex-web-sa':4, 'echoservice':5]
+        },
```
Comment (chhex) [inline common/portnr.gradle]: This clojure envDesignators returns a Map of Target System Environment Character and the corresponding number in the Portnummer.
Open: other environments like Production

File: common/portnr.gradle (hunk 2 above)
Comment (chhex) [reply common/portnr.gradle]: Below, the clojure serviceNames, which returns the a Map of serviceName as Key and Portnummer Part as Value. New services are configured here

File: common/portnr.gradle (hunk 7 above)
Comment (apgsga-uge) [reply common/portnr.gradle]: question here: do we support the quality ensurance environment CHQI211? then we'd need also **'Q':2**

Comment (chhex) [reply common/portnr.gradle]: Yep, and also the 'P'.

**Comment by apgsga-jhe** [comment]
I like the approach of having the logic and the config within the same gradle file.

In general, I would throw exceptions when anything is missing in the config (a service name, a target, etc ...), I wouldn't install services with default port if it wasn't correctly configured.

Regarding the place where to put this code ... Nice thing when having it within this project is that we can easily update the configuration. On the other side, this doesn't need to be available for ALL Gradle projects, but it doesn't hurt and no one could do something bad by using this script. Having said that, I think I would keep it here for now, my main argument being that it's easy to change it, and easy to track the changes (git ...).

Regarding "The calculation of additional Ports for a Service" ... Yep, to be discussed.

**Pull request merged**